### ISSUE TITLE
Add an easier way to create memory references that point to a different block

### ISSUE THREAD

**Issue by agatti** [NationalSecurityAgency/ghidra#6268]
**Is your feature request related to a problem? Please describe.**

When analysing executable with more than one code segment (ie. bank-switched firmware, real mode DOS executables, etc.), memory offsets may be for a different block than the one they are found in.  In that case, creating a correct reference takes a lot of steps (open the references editor, bring up the add reference dialog, click on the memory radio button, calculate the new address by hand, type it in hoping there are no typos, click add).  If automatic analysis couldn't figure out the proper block (or worse, forcing references to the wrong block) and you've got more than a handful of those addresses it gets annoying real quick.

**Describe the solution you'd like**

IDA Pro has a simpler way to tackle this, by having a shortcut to choose the block the current offset refers to (Alt+R - "Offset by (any segment)..."), using the selected value as a displacement into the chosen block.  Something similar that optionally reuses the last chosen block would be nice to have (say, Ctrl+Alt+&lt;something&gt; for the selection dialog, Ctrl+Shift+Alt+&lt;something&gt; for forcing the last chosen block).

**Describe alternatives you've considered**

I've hacked together a GhidraScript that sort of does this but I think the functionality should be built-in in Ghidra (my script is just some code copied from the Memory Map and References plugins and won't really support all possible cases).  Having it as a script brings up three windows each time the script is invoked (block selection, console, and cancellation dialog).

**Additional context**

The GhidraScript file in question can be found here: https://gist.github.com/agatti/ccb724bc33f8a63f9003dccaefeae8a8

**Comment by emteere** [comment]
I agree this would be a useful feature.   Will take a look at what you've done.

The last type of address created, say an offset from an address (the beginning of your segment) could be used in a repeat action that would create the same type of address with the new offset.

References can be created manually to other memory locations (overlays, other spaces, etc...).  But the decompiler will only use the reference in certain cases.  For example, an overridden call target address will get used as the call target by the decompiler.

A Pointer Typedef can be used on a parameter or memory location to automatically get a constant value to be coerced to a memory location in an alternate address space.  The pointer typedef is a typedef on a pointer with settings on it for the alternate address space.  I can help the decompiler use the value passed as a parameter correctly as a pointer to the alternate address space.

You can use a pointer typedef to refer to another address space, say data instead of ram on a harvard architecture, or you can define new named address spaces as a named overlay space, or artificial other space. 

This works well you can find a place to hang the data type that the decompiler will use to interpret the value.
It doesn't work in all cases, for example if another register value is needed to complete the pointer.  There are plans to make it work for segment registers or other values that need more information to correctly use the value as a pointer.

They are also a bit hidden and hard to create.
- Create a pointer to a data type in the data type manager
- The create a new Typedef on the pointer to the data type in the data type manager
- Then edit the settings of the new Pointer Typedef

**Comment by Wall-AF** [comment]
Better segment support is definately required.  X86 _can_ use them even in 64-bit applications (although not so much in standard applications).

**Comment by pcjc2** [comment]
I believe this would be very useful for analyzing embedded code, where different operation modes employ a finite RAM resource (or similar) for different global structures.
